AWSS3/AWSS3TransferUtility.m

```diff
@@ -2226,7 +2233,18 @@ - (void)URLSession:(NSURLSession *)session
             subTask.eTag = (NSString *) HTTPResponse.allHeaderFields[@"ETAG"];
             
             //Add it to completed parts and remove it from remaining parts.
-            [transferUtilityMultiPartUploadTask.completedPartsDictionary setObject:subTask forKey:@(subTask.taskIdentifier)];
+
+            // Don't clobber existing keys...
+            if ([transferUtilityMultiPartUploadTask.completedPartsDictionary objectForKey:@(subTask.taskIdentifier)]) {
```

Comment:
The change looks like it will fix the issue, however my concern is for future readers to understand why there would be an existing key in the `completedPartsDictionary` without going over this PR to understand that taskIdentifiers from NSURLSessionUploadTask is not unique in some scenarios. 

My question is then can we use the `subtask.partNumber` to uniquely identify the subtasks? This seems like a logical change at first and would apply to updating the structures of `waitingPartsDictionary`, `completedPartsDictionary`, and `inProgressPartsDictionary` but I'm afraid of a regression without diving into every interaction of those three structures.

So I think a more isolated change, that will still keep the door open for possibly using partNumber as the key for `inProgressPartsDictionary` and `waitingPartsDictionary` is to only change the strcture of `completedPartsDictionary` to
```
@property (strong, nonatomic) NSMutableSet <AWSS3TransferUtilityUploadSubTask *> *completedPartsSet;
```

Running over some of the uses of the set

Hydrating from DB should add it to the set uniquely based on the object which includes the taskIdentifier,  partNumber, and rest of the fields
```objective-c
//Check if the subTask is is already completed. If it is, add it to the completed parts list, update the progress object and go to the next iteration of the loop
            if (subTask.status== AWSS3TransferUtilityTransferStatusCompleted ) {
                [multiPartUploadTask.completedPartsSet addObject:subTask];
                continue;
            }
```

Calling finish will construct the AWSS3CompletedPart in ascending order of partNumber
```objective-c
//Create a new Dictionary with the partNumber as the Key
    for(AWSS3TransferUtilityUploadSubTask *subTask in uploadTask.completedPartsSet) {
        [tempDictionary setObject:subTask forKey:subTask.partNumber];
    }
    
    //Compose the request.
    for(int i = 1; i <= [uploadTask.completedPartsSet count]; i++) {
        AWSS3TransferUtilityUploadSubTask *subTask = [tempDictionary objectForKey: [NSNumber numberWithInt:i]];
        AWSS3CompletedPart *completedPart = [AWSS3CompletedPart new];
        completedPart.partNumber = subTask.partNumber;
        completedPart.ETag = subTask.eTag;
        [completedParts addObject:completedPart];
    }
```

OnComplete (the code at this line) will just add it to the set and be deduped to a new element in the set even when the taskIdentifer is the same (assuming partNumber if the unique identifier)
```objective-c
//Add it to completed parts and remove it from remaining parts.
            [transferUtilityMultiPartUploadTask.completedPartsSet addObject:subTask];
            [transferUtilityMultiPartUploadTask.inProgressPartsDictionary removeObjectForKey:@(subTask.taskIdentifier)];
```

Reply:
Will wait for your feedback on my comments regarding the use of a Set instead of generating a random number, here's the PR with my proposed change https://github.com/aws-amplify/aws-sdk-ios/pull/3282/files Since I couldn't reproduce  yet, wanted to still send this out incase you get a chance verify the fix on your end